### PR TITLE
stft: remove non-center overload and python functional wrapper

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73434

### CODE REVIEW

File: torch/functional.py [hunk 1]
```diff
@@ -476,133 +476,121 @@ def _meshgrid(*tensors, indexing: Optional[str]):
...
+    return_complex (bool, optional): whether to return a complex tensor, or
+        a real tensor with an extra last dimension for the real and
+        imaginary components.
+
+Returns:
+    Tensor: A tensor containing the STFT result with shape described above
+
+""")
+# TODO: Fix via https://github.com/pytorch/pytorch/issues/75798
+stft.__module__ = "torch.functional"
```
Comment (anjali411) [inline torch/functional.py:593]: shouldn't be needed anymore

**Review by anjali411 — APPROVED** [review]
LGTM! @peterbell10 the FC period is over right?

**Comment by peterbell10** [comment]
Hrm, that's unfortunate. Well for whoever does have the right permissions, this stack needs to be reverted because an xfail on a slow test is now passing due to the fx bug-fix.

**Comment by zengk95** [comment]
@pytorchbot revert this

This is breaking slow tests.

**Comment by peterbell10** [comment]
Oh actually, the fix was just merged in #76268.

**Comment by seemethere** [comment]
> Oh actually, the fix was just merged in #76268.

In the future we should prefer to revert rather than forward fix. Also we should try to avoid landing code late at the end of the week to avoid our oncalls having to potentially revert things on the weekend

**Comment by osalpekar** [comment]
@pytorchbot revert this please

This broke functorch - error log below:
```
Summary: 
pytorch/functorch/functorch/csrc/VmapGeneratedPlumbing.h:10831:114: error: too few arguments to function call, expected 10, have 8
    return at::_ops::stft::call(self, n_fft, hop_length, win_length, window, normalized, onesided, return_complex);
           ~~~~~~~~~~~~~~~~~~~~                                                                                  ^
pytorch/functorch/functorch/csrc/VmapGeneratedPlumbing.h:10851:22: error: no member named 'stft_center' in namespace 'at::_ops'
    return at::_ops::stft_center::call(self, n_fft, hop_length, win_length, window, center, pad_mode, normalized, onesided, return_complex);
           ~~~~~~~~~~^
pytorch/functorch/functorch/csrc/VmapGeneratedPlumbing.h:10831:114: error: too few arguments to function call, expected 10, have 8
    return at::_ops::stft::call(self, n_fft, hop_length, win_length, window, normalized, onesided, return_complex);
           ~~~~~~~~~~~~~~~~~~~~                                                                                  ^
pytorch/functorch/functorch/csrc/VmapGeneratedPlumbing.h:10851:22: error: no member named 'stft_center' in namespace 'at::_ops'
    return at::_ops::stft_center::call(self, n_fft, hop_length, win_length, window, center, pad_mode, normalized, onesided, return_complex);
           ~~~~~~~~~~^
```
cc @zou3519 @Chillee

**Comment by zou3519** [comment]
I discussed offline with @peterbell10 -- I'll handle importing this to Phabricator and landing it to avoid the functorch failure

**Comment by zou3519** [comment]
We delete stft from functorch, so this PR should be good when merged the usual way.

**Comment by peterbell10** [comment]
>This PR seems to be causing slight forward-backward compatibility problems (by removing `torch.functional.F` symbol)

You mean someone is explicitly importing `torch.functional.F` instead of `torch.nn.functional`? By that logic is every import removal considered a BC-break?

>can you please provide a more detailed description to your PRs especially if they introduce something that looks like backward incompatible changes.

I thought the whole point of the JIT upgraders is that this isn't considered a BC break any more? Or do you mean the `import`?

**Comment by anjali411** [comment]
Hey @malfet `F` should have never been exposed in `torch.functional` so this is also goes back to a bigger discussion of our public API policies. Internal failures were fixed by Natalia (see https://fb.workplace.com/groups/1405155842844877/posts/5840766845950399/?comment_id=5841992042494546&reply_comment_id=5842183229142094)

cc. @ngimel @albanD

**Comment by malfet** [comment]
Hi @anjali411 , there were few more internal projects that dependent on the change, so I have to push some updates as well.
But it looks like upgrader is not configured correctly, as running something as simple as:
```
$ python -c "import torch;import torchaudio; torch.jit.script(torchaudio.transforms.Spectrogram()).save('foo.pth')"
```
before the change and the loading the model after results in an error:
```
% python -c "import torch;print(torch.load('foo.pth'))"                                                            
/Users/nshulga/miniconda3/envs/py_39/lib/python3.9/site-packages/torch/serialization.py:707: UserWarning: 'torch.load' received a zip file that looks like a TorchScript archive dispatching to 'torch.jit.load' (call 'torch.jit.load' directly to silence this warning)
  warnings.warn("'torch.load' received a zip file that looks like a TorchScript archive"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/nshulga/miniconda3/envs/py_39/lib/python3.9/site-packages/torch/serialization.py", line 711, in load
    return torch.jit.load(opened_file)
  File "/Users/nshulga/miniconda3/envs/py_39/lib/python3.9/site-packages/torch/jit/_serialization.py", line 164, in load
    cpp_module = torch._C.import_ir_module_from_buffer(
RuntimeError: 

aten::stft(Tensor self, int n_fft, int? hop_length=None, int? win_length=None, Tensor? window=None, bool center=True, str pad_mode="reflect", bool normalized=False, bool? onesided=None, bool? return_complex=None) -> (Tensor):
Expected a value of type 'str' for argument 'pad_mode' but instead found type 'Optional[bool]'.
:
  File "code/__torch__/torch/functional.py", line 21
  else:
    input0 = input
  _2 = torch.stft(input0, n_fft, hop_length, win_length, window, normalized, onesided, return_complex)
       ~~~~~~~~~~ <--- HERE
  return _2
'stft' is being compiled since it was called from 'spectrogram'
  File "/Users/nshulga/miniconda3/envs/py_3.7-torch-1.11/lib/python3.7/site-packages/torchaudio/functional/functional.py", line 102

    # default values are consistent with librosa.core.spectrum._spectrogram
    spec_f = torch.stft(
             ~~~~~~~~~~ <--- HERE
        input=waveform,
        n_fft=n_fft,
Serialized   File "code/__torch__/torchaudio/functional/functional.py", line 25
  shape = torch.size(waveform0)
  waveform2 = torch.reshape(waveform0, [-1, shape[-1]])
  spec_f = __torch__.torch.functional.stft(waveform2, n_fft, hop_length, win_length, window, center, pad_mode, False, onesided, True, )
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
  _1 = torch.add(torch.slice(shape, None, -1), torch.slice(torch.size(spec_f), -2))
  spec_f0 = torch.reshape(spec_f, _1)
'spectrogram' is being compiled since it was called from 'Spectrogram.forward'
Serialized   File "code/__torch__/torchaudio/transforms.py", line 18
  def forward(self: __torch__.torchaudio.transforms.Spectrogram,
    waveform: Tensor) -> Tensor:
    _0 = __torch__.torchaudio.functional.functional.spectrogram
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    window = self.window
    center = self.center
```
Which also affects internal pipeline, see https://fb.workplace.com/groups/4571909969591489/posts/5035659103216571/?comment_id=5036153876500427

**Comment by peterbell10** [comment]
Can someone revert?

**Comment by albanD** [comment]
@pytorchbot revert this please as it breaks backward compatibility of torchaudio models

Doing this for Peter above who doesn't have the rights.

**Comment by vmoens** [comment]
FYI This PR also seems to be [BC breaking on some torch hub models](https://github.com/pytorch/hub/issues/283)

**Comment by peterbell10** [comment]
Can't reopen this PR, so have opened #77257. Will continue discussion there.

**Pull request closed without merging**